File: py-algorithm/incomplete/strip_comment.py
import unittest

"""
    https://www.codewars.com/kata/51c8e37cee245da6b40000bd/train/python/
    
    Xóa commnet trong chuỗi:
    
        apples, pears # and bananas         apples, pears
        grapes                        ->    grapes
        bananas !apples                     bananas
"""
class Solution(unittest.TestCase):

    @staticmethod
    def strip_comments(text, markers):
        ss = text.split('\n')
        for i, s in enumerate(ss):
            for marker in markers:
                index = s.find(marker)
                if index != -1:
                    s = s[:index].rstrip()
            ss[i] = s

        return '\n'.join(ss)

    def test(self):
        self.assertEqual("apples, pears\ngrapes\nbananas",
                         self.strip_comments("apples, pears # and bananas\ngrapes\nbananas !apples", ["#", "!"]))
        self.assertEqual('a\nc\nd', self.strip_comments('a #b\nc\nd $e f g', ['#', '$']))
        self.assertEqual(' a\nc\nd', self.strip_comments(' a #b\nc\nd $e f g', ['#', '$']))


if __name__ == '__main__':
    unittest.main()

File: codewars/src/main/java/string/StripComment.java
package string;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

import org.apache.commons.lang3.StringUtils;

/**
 * https://www.codewars.com/kata/51c8e37cee245da6b40000bd/
 */
public class StripComment {

    public static String stripComments(String text, String[] markers) {
        String[] ss = text.split("\n");
        for (int i = 0; i < ss.length; i++) {
            String s = ss[i];
            for (String marker : markers) {
                int index = s.indexOf(marker);
                if (index != -1) {
                    s = s.substring(0, index);

                }
            }
            s = StringUtils.stripEnd(s, null);
            ss[i] = s;
        }
        return String.join("\n", ss);
    }

    @Test
    public void stripComments() {
        assertEquals(
                "apples, pears\ngrapes\nbananas",
                stripComments("apples, pears # and bananas\ngrapes\nbananas !apples", new String[]{"#", "!"})
        );

        assertEquals(
                "a\nc\nd",
                stripComments("a #b\nc\nd $e f g", new String[]{"#", "$"})
        );
    }
}
